### DISCUSSION TITLE
Is there a way to let the model have access to the current time in case I ask what time is it?

### DISCUSSION THREAD

**Discussion by joseiriarte1982** [ggml-org/llama.cpp#4730] [Q&A]
The thing is I am running a model from Hugging face dolphin mistral but it doesn't have access to current time, how could I go on solving my issue so it has access to current time when running on chat or -cml, thanks in advance

**Comment by skidd-level-100** [comment]
I am actually working on something with that same model
it can do much more than time.
https://github.com/ggerganov/llama.cpp/discussions/4731

I have not released code yet but when its not SO UGLY of code I will.

**Comment by bullno1** [comment]
From easiest to hardest:

# System prompt

Put the current date in the system prompt. For time, it will suck since you have to re evaluate.
It works well enough if you only care about date.

# The message before

Put the current date right before the current message and then rollback the context.
This still invalidates the KV cache but not as much.
For example, if the user ask: "How long is it until Christmas":

> User: Previous question
> Agent; Previous answer
> User: What is the current time?
> Agent: { Insert the current date here }
> User: How long is it until Christmas
> Agent: { Start generation }

You'd have to remove the date Q&A pair for every message.

I believe with the new cache API, you can shift the "How long is it until Christmas?" pair backward and overwrite the Q&A pair about the current date so there might not be a performance penalty but I haven't tried.

# Special syntax

Follow the [GSM8K](https://github.com/openai/grade-school-math) footstep and make it ask for the date with special syntax.
Then stop sampling, inject the current date and continue.

I find that in-context few-shot prompting works really well even with 7b models.
You don't even need to finetune for it.
